acceptancetests/assess_network_spaces.py

```diff
@@ -0,0 +1,600 @@
+        if target_model:
+            self.connect_to_existing_model(client, target_model)
+        else:
+            self.assign_spaces(client)
+        # add machines for spaces testing
+        # (should we do this if we are on an existing model?)
+        self.deploy_spaces_machines(client, series)
+
+
+    def connect_to_existing_model(self, client, target_model):
+        """Connects to an existing Juju model.
+
+        :param client: Juju client object without bootstrapped controller
+        :param target_model: Model to connect to for testing
+        """
+        log.info("Connecting to existing model: {}".format(target_model))
+        if client.show_model().keys()[0] is not target_model:
+            client.switch(target_model)
+
+
+    def assign_spaces(self, client):
+        """Assigns spaces to subnets
+        Name the spaces sequentially: space1, space2, space3, etc.
+        We require at least 3 spaces.
+
+        :param client: Juju client object with controller
+        """
+        log.info('Assigning network spaces on {}.'.format(client.env.provider))
+        subnets = yaml.safe_load(client.get_juju_output('list-subnets',
+            '--format=yaml'))
+        if not subnets:
+            raise SubnetsNotReady(
+                'No subnets defined in {}'.format(client.env.provider))
+        subnet_count = 0
+        for subnet in non_infan_subnets(subnets)['subnets'].keys():
+            subnet_count += 1
+            client.juju('add-space', ('space{}'.format(subnet_count), subnet))
+        if subnet_count < 3:
+            raise SubnetsNotReady('3 subnets required for spaces assignment. '
+                            '{} found.'.format(subnet_count))
+
+
+    def verify_machine_spaces(self, client):
+        """Check all the machines to verify they are in the expected spaces
+        We should have 4 machines in 3 spaces
+        0 and 1 in space1
+        2 in space2
+        3 in space3
+
+        :param client: Juju client object with machines and spaces
+        :returns pass: true/false upon pass/fail
+        :returns message: failure message
+        """
+        log.info('Assessing machines are in the correct spaces.')
+        spaces = non_infan_subnets(
+            yaml.safe_load(
+                client.get_juju_output(
+                    'list-spaces', '--format=yaml')))
+        machines = yaml.safe_load(
+            client.get_juju_output(
+                'list-machines', '--format=yaml'))['machines']
+
+        results = {
+            'pass': True,
+            'message': ''
+        }
+        for machine in machines.keys():
+            if machine == '0':
+                expected_space = 'space1'
+            else:
+                expected_space = 'space{}'.format(machine)
+            eth0 = machines[machine]['network-interfaces']['eth0']
+            subnet = spaces['spaces'][expected_space].keys()[0]
+            for ip in eth0['ip-addresses']:
```

Comment:
If I'm reading this right, than any IP in eth0['ip-addresses'] that isn't in the subnet cidr is a failure for the test? (and the successful ones don't really matter?)

This could be simplified (and IMO easier to read) as something like:
```python
ip_address = eth0['ip-addresses']
ips_in_correct_space = [ip for ip in ip_addresses if ip_in_cidr(ip, subnet)]
missing_ips = set(ip_addresses).difference(ips_in_correct_space)
if missing_ips != set():
    message = 'Machine {} IPs {} where not in ...'.format(', '.join(missing_ips))
```

Reply:
It should be the other way around: As long as we have one IP on eth0 in the correct space, we're good. I see now that my logic fails if the good IP is not first.

Reply:
Ah right, that simplifies it even further, it can be expressed as:

```python
if not any([ip for ip in ip_addresses if ip_in_cidr(ip, subnet)]):
    raise TestError('None of the IPs where in the CIDR . . .')
```